README.md
# identisea

Brute force search for mathematical identities

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar identisea-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2015 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2015-12-11
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2015-12-11
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/identisea/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/identisea/compare/0.1.0...0.1.1


test/identisea/core_test.clj
(ns identisea.core-test
  (:require [clojure.test :refer :all]
            [identisea.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/identisea/meshed.clj
(ns identisea.meshed
  (:gen-class)
  (:require [clojure.math.numeric-tower :as math]
            [clojure.math.combinatorics :as comb]
            [serializable.fn :as sfn]))

(defn result-mesh
  [functions value-sets]
  (for [f functions]
    (with-meta 
      (map (partial apply f) value-sets) 
      (meta f))))

(def add
  (with-meta
    (sfn/fn [x y] (+ x y))
    {:func "add"}))

(def mul
  (with-meta
    (sfn/fn [x y] (* x y))
    {:func "mult"}))

(def exp
  (with-meta 
    (sfn/fn [x y] (math/expt x y))
    {:func "exp"}))

(def modulo
  (with-meta
    (sfn/fn [x y] (mod x y))
    {:func "mod"}))

(def fermat
  (with-meta 
    (sfn/fn [x y] (modulo (exp x y) y))
    {:func "fermat"}))

(def functions [add mul exp modulo fermat])

(def const-min 2)
(def const-max 50)

(def functions
  (into functions 
    (for [c (range const-min const-max)]
      (sfn/fn [x &xs] c))))

(def search-size 500)

(def value-sets
  (for [x (range 2 search-size)
        y (range 2 search-size)]
    [x y]))

(defn decimal-true
  [values]
  (let [cutoff (/ 100.0 search-size)
        valid (count (filter identity values))
        result (* 100 (float (/ valid (count values))))]
    (if (> result cutoff)
      result
      0.0)))

(defn search-mesh
  [mesh comparison]
    (for [[f1 f2] (comb/combinations mesh 2)]
      (with-meta [(decimal-true (map comparison f1 f2))]
                 {:f1 (:func (meta f1)) :f2 (:func (meta f2)) :comp "="})))

(defn format-function-pair-result
  [meta-val]
    (str meta-val " " (:f1 (meta meta-val)) "-" (:f2 (meta meta-val))))

(defn main []
  (let [rm (result-mesh functions value-sets)]
    (doall (map println 
                (map format-function-pair-result
                     (search-mesh rm =))))))





src/identisea/reexpander.clj
(ns identisea.reexpander
  (:gen-class)
  (:require [clojure.math.numeric-tower :as math]
            [clojure.math.combinatorics :as comb]
            [serializable.fn :as sfn]))

(defn result-mesh
  [functions value-sets]
  (for [f functions]
    (with-meta 
      (map (partial apply f) value-sets) 
      (meta f))))

(def variables 2)

(defprotocol Operation
  (operate [this t p])
  (last-op [this]))

(def r-add
  (for [i (range variables)]
    (reify Operation
      (operate [this t p] (+ t (nth p i)))
      (last-op [this] "r-add"))))

(def varbase 
  (for [i (range variables)]
    (with-meta (fn [p] (nth p i))
               {:math :var :str (str "#" i)})))

; should use a macro to create operators 
(def t-add 
  (with-meta (fn [i o] 
               (fn [p] (+ (i p) (o p)))) 
             {:math :lin :fmt "(+ %1$s %2$s)"}))

(def t-multiply 
  (with-meta (fn [i o] 
               (fn [p] (* (i p) (o p)))) 
             {:math :geo :fmt "(* %1$s %2$s)"}))

(def t-exp
  (with-meta (fn [i o] 
               (fn [p] (math/expt (i p) (o p)))) 
             {:math :exp :fmt "(exp %1$s %2$s)"}))

(def exp-t
  (with-meta (fn [i o] 
               (fn [p] (math/expt (o p) (i p)))) 
             {:math :exp :fmt "(exp %1$s %2$s)"}))

(def t-mod
  (with-meta (fn [i o] 
               (fn [p] (mod (i p) (o p)))) 
             {:math :mod :fmt "(mod %1$s %2$s)"}))

(def mod-t
  (with-meta (fn [i o] 
               (fn [p] (mod (o p) (i p)))) 
             {:math :mod :fmt "(mod %1$s %2$s)"}))

#_(def fermat
  (with-meta 
    (fn [x y] (x-mod-y (x-exp-y x y) y))
    {:func "fermat"}))

(def functions varbase)

(def operations [t-add t-multiply t-exp t-mod])

#_(def functions
  (map (fn [f] (vary-meta f assoc :last (:math (meta f)))) functions))

(def complexity 2)
(def const-min 2)
(def const-max 50)

(defn non-redundant-composition 
  [o f1 f2]
  (and (not= (:math (meta o)) (:math (meta f1)))
       (not= (:math (meta o)) (:math (meta f2)))))

(defn compose-functions 
  [o f1 f2]
  (let [mo (meta o)
        srep (format (:fmt mo) (:str (meta f1)) (:str (meta f2)))]
    (with-meta (o f1 f2) (assoc mo :str srep)))) 

(doseq [i (range complexity)]
  (def functions
    (into functions
      (for [o operations
            f1 functions
            f2 functions
            :when (non-redundant-composition o f1 f2)] 
        (compose-functions o f1 f2)))))

(doseq [f functions]
  (println (:str (meta f))))

(def search-size 10)

(defn decimal-true
  [values]
  (let [cutoff (/ 100.0 search-size)
        valid (count (filter identity values))
        result (* 100 (float (/ valid (count values))))]
    (if (> result cutoff)
      result
      0.0)))

(defn format-equality
  [f1 f2]
  (format "%s = %s" (:str (meta f1)) (:str (meta f2))))

(defn search-mesh
  [mesh comparison]
    (for [[f1 f2] (comb/combinations mesh 2)]
      (with-meta [(decimal-true (map comparison f1 f2))]
                 {:str (format-equality f1 f2)})))

(defn format-function-pair-result
  [meta-val]
    (str meta-val " " (:str (meta meta-val)) "-" (:str (meta meta-val))))


(def mesh
  (for [p (comb/combinations (range 2 search-size) variables)]
    (for [f functions]
      (f p))))

(defn main []
  (doall (map #(println (str % " " (:str (meta %)))) 
              (search-mesh mesh =))))




src/identisea/expander.clj
(ns identisea.expander
  (:gen-class)
  (:require [clojure.math.numeric-tower :as math]
            [clojure.math.combinatorics :as comb]
            [serializable.fn :as sfn]))

(defn result-mesh
  [functions value-sets]
  (for [f functions]
    (with-meta 
      (map (partial apply f) value-sets) 
      (meta f))))

(def variables 2)

(defprotocol Operation
  (operate [this t p])
  (last-op [this]))

(def r-add
  (for [i (range variables)]
    (reify Operation
      (operate [this t p] (+ t (nth p i)))
      (last-op [this] "r-add"))))

(def const
  (for [i (range variables)]
    (with-meta (fn [p] (nth p i))
               {:last :const :var i})))

(def t-add
  (for [i (range variables)] 
    (with-meta (fn [t p] (+ t (nth p i))) 
               {:last :add :var i})))

(def t-multiply
  (for [i (range variables)]
    (with-meta (fn [t p] (* t (nth p i)))
               {:last :mult :var i})))

(def t-exp
  (for [i (range variables)]
    (with-meta (fn [t p] (math/expt t (nth p i)))
               {:last :t-exp :var i))

(def exp-t
  (for [i (range variables)]
    (with-meta (fn [t p] (math/expt (nth p i) t))
               {:last :exp-t :var i})))

(def t-mod
  (for [i (range variables)]
    (with-meta (fn [t p] (mod t (nth p i)))
               {:last :mod-t :var i})))

(def mod-t
  (for [i (range variables)]
    (with-meta (fn [t p] (mod (nth p i) t))
               {:last :t-mod :var i})))

(doseq [f t-add]
  (println f))


(def x-add-y 
  (with-meta
    (sfn/fn [x y] (+ x y))
    {:func "x-add-y"}))

(def x-mult-y 
  (with-meta
    (sfn/fn [x y] (* x y))
    {:func "x-mult-y"}))

(def x-exp-y
  (with-meta 
    (sfn/fn [x y] (math/expt x y))
    {:func "x-exp-y"}))

(def x-mod-y 
  (with-meta
    (sfn/fn [x y] (mod x y))
    {:func "x-mod-y"}))

(def fermat
  (with-meta 
    (sfn/fn [x y] (x-mod-y (x-exp-y x y) y))
    {:func "fermat"}))

(def functions [x-add-y x-mult-y x-exp-y x-mod-y])
(def functions
  (map (fn [f] (vary-meta f assoc :last (:func (meta f)))) functions))
(def functions
  (map (fn [f] (vary-meta f assoc :path (:func (meta f)))) functions))

(def complexity 2)
(def const-min 2)
(def const-max 50)

(doseq [i (range complexity)]
  (def functions
    (into functions
      (for [f1 functions
            f2 functions
            :when (not= (:last (meta f1)) (:last (meta f2)))] 
        (with-meta (comp f1 f2) 
                   (assoc (meta f1) 
                          :path (str (:path (meta f1)) ":" 
                                     (:path (meta f2) ))))))))

(doseq [f functions]
  (println f))

(def search-size 10)

(def value-sets
  (for [a (range 2 search-size)
        b (range 2 search-size)]
    [(fn [x y] a) 
     (fn [x y] b)]))

(defn decimal-true
  [values]
  (let [cutoff (/ 100.0 search-size)
        valid (count (filter identity values))
        result (* 100 (float (/ valid (count values))))]
    (if (> result cutoff)
      result
      0.0)))

(defn search-mesh
  [mesh comparison]
    (for [[f1 f2] (comb/combinations mesh 2)]
      (with-meta [(decimal-true (map comparison f1 f2))]
                 {:f1 (:func (meta f1)) :f2 (:func (meta f2)) :comp "="})))

(defn format-function-pair-result
  [meta-val]
    (str meta-val " " (:f1 (meta meta-val)) "-" (:f2 (meta meta-val))))

(defn main []
  (let [rm (result-mesh functions value-sets)]
    (doall (map println 
                (map format-function-pair-result
                     (search-mesh rm =))))))



